pkg/controller/hibernation/alibabacloud_actuator_test.go

```diff
@@ -0,0 +1,274 @@
+			testFunc:          "MachinesRunning",
+			expectedResult:    false,
+			expectedRemaining: []string{"Stopping-0", "Stopping-1", "Stopping-2", "Stopped-0"},
+			instances:         map[string]int{"Running": 3, "Stopped": 1, "Stopping": 3},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			alibabaCloudClient := mockalibabacloud.NewMockAPI(ctrl)
+			setupAlibabaCloudClientInstances(alibabaCloudClient, test.instances)
+			if test.setupClient != nil {
+				test.setupClient(t, alibabaCloudClient)
+			}
+			actuator := testAlibabaCloudActuator(alibabaCloudClient)
+			var err error
+			var result bool
+			var remaining []string
+			switch test.testFunc {
+			case "MachinesStopped":
+				result, remaining, err = actuator.MachinesStopped(testAlibabaCloudClusterDeployment(), nil, log.New())
+			case "MachinesRunning":
+				result, remaining, err = actuator.MachinesRunning(testAlibabaCloudClusterDeployment(), nil, log.New())
+			default:
+				t.Fatal("Invalid function to test")
+			}
+			require.Nil(t, err)
+			assert.Equal(t, test.expectedResult, result)
+			if len(test.expectedRemaining) > 0 {
+				sort.Strings(test.expectedRemaining)
+				sort.Strings(remaining)
+				assert.Equal(t, test.expectedRemaining, remaining)
+			}
+		})
+	}
+}
+
+func testAlibabaCloudActuator(alibabaCloudClient alibabaclient.API) *alibabaCloudActuator {
+	return &alibabaCloudActuator{
+		alibabaCloudClientFn: func(*hivev1.ClusterDeployment, client.Client, log.FieldLogger) (alibabaclient.API, error) {
+			return alibabaCloudClient, nil
+		},
+	}
+}
+
+func setupAlibabaCloudClientInstances(alibabaCloudClient *mockalibabacloud.MockAPI, statuses map[string]int) {
+	instances := []ecs.Instance{}
+	for status, count := range statuses {
+		for i := 0; i < count; i++ {
+			instances = append(instances, ecs.Instance{
+				InstanceName: fmt.Sprintf("%s-%d", status, i),
+				InstanceId:   fmt.Sprintf("%s-%d", status, i),
+				Status:       status,
+			})
+		}
+	}
+
+	response := &ecs.DescribeInstancesResponse{
+		BaseResponse: &responses.BaseResponse{},
+	}
+	response.Instances.Instance = instances
+
+	alibabaCloudClient.EXPECT().DescribeInstances(gomock.Any()).Times(1).Return(response, nil)
+}
+
+func testAlibabaCloudClusterDeployment() *hivev1.ClusterDeployment {
+	cdBuilder := testcd.FullBuilder("testns", "testalibabacluster", scheme.Scheme)
+	return cdBuilder.Build(
+		testcd.WithAlibabaCloudPlatform(&hivev1alibabacloud.Platform{Region: "cn-hangzhou"}),
+		testcd.WithClusterMetadata(&hivev1.ClusterMetadata{InfraID: "testalibabacluster-foobarbaz"}),
+	)
+}
+
+func matchAlibabaInstanceIDs(t *testing.T, actual *[]string, states map[string]int) {
+	expected := sets.NewString()
+	for state, count := range states {
+		for i := 0; i < count; i++ {
+			expected.Insert(fmt.Sprintf("%s-%d", state, i))
+		}
+	}
+	actualSet := sets.NewString()
+	for _, a := range *actual {
+		actualSet.Insert(aws.StringValue(&a))
```

Comment:
Could we replace this with `pointer.StringPtr` to keep the AWS lib away from Alibaba? They're probably functionally equivalent it just seems weird to me :laughing:

Reply:
I misread this as pointer...

Reply:
fixed